"expected *DFloat, found *tree.DInt" SQL panic in st_distance()

**Describe the problem**

A panic in the `st_distance()` function. 

**To Reproduce**

(I can provide steps if the log below isn't enough to determine the issue.)

**Additional data / screenshots**
```
E200914 00:47:16.514355 1896 sql/conn_executor.go:803  [n1,client=127.0.0.1:47278,hostnossl,user=root] a SQL panic has occurred while executing the following statement:
SELECT distapp_southtexascity.id, distapp_southtexascity.name, distapp_southtexascity.point::BYTES, distapp_southtexascity.radius FROM distapp_southtexascity WHERE st_distance(distapp_southtexascity.point, st_geomfromewkb('\x01010000208c7d0000b0ffffffc9062d41c1ffff7f20125041')) <= distapp_southtexascity.radius ORDER BY distapp_southtexascity.name ASC
E200914 00:47:16.514478 1896 util/log/crash_reporting.go:138  [n1,client=127.0.0.1:47278,hostnossl,user=root] a panic has occurred!
expected *DFloat, found *tree.DInt
(1) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:508
  | runtime.gopanic
  | 	/usr/local/go/src/runtime/panic.go:679
  | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MustBeDFloat
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:768
  | github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.makeSTDWithinBuiltin.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/geo_builtins.go:5473
  | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3873
  | github.com/cockroachdb/cockroach/pkg/sql/schemaexpr.RunFilter
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/schemaexpr/expr_filter.go:21
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*ExprHelper).EvalFilter
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:189
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcOutputHelper).ProcessRow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:288
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).ProcessRowHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:731
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*tableReader).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:256
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sortAllProcessor).fill
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sorter.go:291
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sortAllProcessor).Start
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sorter.go:272
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:762
  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:380
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:991
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1059
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:930
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:610
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:116
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1459
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1388
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:510
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:626
Wraps: (2) while executing: SELECT _._, _._, _._::BYTES, _._ FROM _ WHERE st_distance(_._, st_geomfromewkb(_)) <= _._ ORDER BY _._ ASC
Wraps: (3) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:508
  | runtime.gopanic
  | 	/usr/local/go/src/runtime/panic.go:679
  | [...repeated from below...]
Wraps: (4) assertion failure
Wraps: (5) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MustBeDFloat
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:768
  | github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.makeSTDWithinBuiltin.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/geo_builtins.go:5473
  | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3873
  | github.com/cockroachdb/cockroach/pkg/sql/schemaexpr.RunFilter
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/schemaexpr/expr_filter.go:21
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*ExprHelper).EvalFilter
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:189
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcOutputHelper).ProcessRow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:288
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).ProcessRowHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:731
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*tableReader).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:256
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sortAllProcessor).fill
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sorter.go:291
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sortAllProcessor).Start
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sorter.go:272
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:762
  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:380
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:991
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1059
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:930
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:610
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:116
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1459
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1388
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:510
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:626
  | runtime.goexit
  | 	/usr/local/go/src/runtime/asm_amd64.s:1357
Wraps: (6) expected *DFloat, found *tree.DInt
Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *assert.withAssertionFailure (5) *withstack.withStack (6) *errutil.leafError
E200914 00:47:16.517162 1896 util/log/crash_reporting.go:321  [n1,client=127.0.0.1:47278,hostnossl,user=root] Queued as error 7789b1d623be4ea5a80c91f902ef4395
panic: expected *DFloat, found *tree.DInt [recovered]
	panic: expected *DFloat, found *tree.DInt

goroutine 1896 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc001ef5000, 0x61698e0, 0xc001e24100, 0x4c92680, 0xc0043ac010)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:820 +0x359
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc001ef5000, 0x61698e0, 0xc001e24100)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:508 +0x61
panic(0x4c92680, 0xc0043ac010)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MustBeDFloat(0x61876a0, 0xc004319200, 0xc004319200)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:768 +0xf3
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.makeSTDWithinBuiltin.func1(0xc0031b3d40, 0xc004c79c50, 0x3, 0x3, 0x3, 0x3, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/geo_builtins.go:5473 +0x151
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0xc003cd3880, 0xc0031b3d40, 0x0, 0xea, 0xc004321ff0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3873 +0xb5
github.com/cockroachdb/cockroach/pkg/sql/schemaexpr.RunFilter(0x61afec0, 0xc003cd3880, 0xc0031b3d40, 0x0, 0x1, 0xc004321ff0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/schemaexpr/expr_filter.go:21 +0x48
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*ExprHelper).EvalFilter(0xc003160c00, 0xc001bd8a80, 0x4, 0x4, 0xc000fe16a0, 0x213783f, 0xc003f29540)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:189 +0xf2
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcOutputHelper).ProcessRow(0xc002ec5018, 0x61698e0, 0xc004b00ac0, 0xc001bd8a80, 0x4, 0x4, 0x4db37e0, 0xc004319188, 0x4, 0x78, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:288 +0x49c
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).ProcessRowHelper(0xc002ec5000, 0xc001bd8a80, 0x4, 0x4, 0x4, 0x4, 0x6315da0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:731 +0x75
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*tableReader).Next(0xc002ec5000, 0x7, 0xc000fe1701, 0x61698e0, 0xc004b00ac0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:256 +0xc9
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sortAllProcessor).fill(0xc0048aa000, 0x61698e0, 0xc004b00ac0, 0x509b8ad)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sorter.go:291 +0x73
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sortAllProcessor).Start(0xc0048aa000, 0x61698e0, 0xc004b00ac0, 0x94bc960, 0x492cf40)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sorter.go:272 +0xab
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run(0xc0048aa000, 0x61698e0, 0xc004b00ac0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:762 +0x52
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc0049f9c20, 0x61698e0, 0xc004b00ac0, 0x5691ce8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:380 +0x24d
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc001267d90, 0xc0042eae00, 0xc004634510, 0xc0040cdd00, 0xc002a5d500, 0xc001ef5510, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422 +0x5d9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc001267d90, 0x61699a0, 0xc004c78cc0, 0xc001ef5510, 0xc0042eae00, 0xc004634510, 0x616c360, 0xc0049926e0, 0x0, 0xc002a5d500, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:991 +0x1d1
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc001ef5000, 0x61699a0, 0xc004c78cc0, 0xc001ef5420, 0x3, 0x7f1fb888c818, 0xc0008caf80, 0xc000fe2201, 0xc002dbcbf8, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1059 +0x3c0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc001ef5000, 0x61699a0, 0xc004c78cc0, 0xc001ef5420, 0x7f1fb888c818, 0xc0008caf80, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:930 +0x6ff
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc001ef5000, 0x61699a0, 0xc004c78cc0, 0x6188720, 0xc004391a40, 0xc0035b06d6, 0x185, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:610 +0xb58
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc001ef5000, 0x61699a0, 0xc004c78cc0, 0x6188720, 0xc004391a40, 0xc0035b06d6, 0x185, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:116 +0x909
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc001ef5000, 0x61698e0, 0xc0031ecdc0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1459 +0x1c77
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc001ef5000, 0x61698e0, 0xc001e24100, 0xc000bd2dc0, 0x5400, 0x15000, 0xc000bd2e60, 0xc001700b30, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1388 +0x1f2
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000c4c100, 0x61698e0, 0xc001e24100, 0xc001ef5000, 0x5400, 0x15000, 0xc000bd2e60, 0xc001700b30, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:510 +0x104
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc001fc98ce, 0xc00178fd20, 0x61698e0, 0xc001e24100, 0xc001700b30, 0xc000c4c100, 0xc0008ca900, 0x61970e0, 0xc00115ba70, 0xc00157e300, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:626 +0x317
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:554 +0x17e
```

**Expected behavior**

No panic.

**Environment:**
 - CockroachDB version 20.2 nightly built 2020/09/13 15:22:30 (this doesn't happen in alpha 3)

--- a/pkg/sql/opt/memo/typing.go
+++ b/pkg/sql/opt/memo/typing.go
@@ -119,6 +119,9 @@ func FindFunction(
 	props, overloads := builtins.GetBuiltinProperties(name)
 	for o := range overloads {
 		overload = &overloads[o]
+		if overload.Types.Length() != e.ChildCount() {
+			continue
+		}
 		matches := true
 		for i, n := 0, e.ChildCount(); i < n; i++ {
 			typ := e.Child(i).(opt.ScalarExpr).DataType()

--- a/pkg/sql/opt/norm/comp_funcs.go
+++ b/pkg/sql/opt/norm/comp_funcs.go
@@ -236,11 +236,14 @@ func (c *CustomFuncs) makeSTDWithin(
 			name = incName
 		}
 	}
-	props, overload, ok := memo.FindFunction(&args, name)
+
+	// The distance parameter must be type float.
+	newArgs := append(args, c.f.ConstructCast(bound, types.Float))
+	props, overload, ok := memo.FindFunction(&newArgs, name)
 	if !ok {
 		panic(errors.AssertionFailedf("could not find overload for %s", name))
 	}
-	within := c.f.ConstructFunction(append(args, bound), &memo.FunctionPrivate{
+	within := c.f.ConstructFunction(newArgs, &memo.FunctionPrivate{
 		Name:       name,
 		Typ:        types.Bool,
 		Properties: props,

--- a/pkg/sql/opt/norm/testdata/rules/comp
+++ b/pkg/sql/opt/norm/testdata/rules/comp
@@ -844,6 +844,37 @@ select
  └── filters
       └── st_dwithinexclusive(geog:2, '0101000020E610000000000000000000000000000000000000', 5.0) [outer=(2), immutable]
 
+# Regression test for #54326. Ensure the distance param is cast to a float.
+norm expect=FoldCmpSTDistanceLeft format=(show-scalars,show-types)
+SELECT * FROM geom_geog WHERE st_distance(geog, 'point(0.0 0.0)') < 5::int
+----
+select
+ ├── columns: geom:1(geometry) geog:2(geography) val:3(float)
+ ├── immutable
+ ├── scan geom_geog
+ │    └── columns: geom:1(geometry) geog:2(geography) val:3(float)
+ └── filters
+      └── function: st_dwithinexclusive [type=bool, outer=(2), immutable]
+           ├── variable: geog:2 [type=geography]
+           ├── const: '0101000020E610000000000000000000000000000000000000' [type=geography]
+           └── const: 5.0 [type=float]
+
+norm expect=FoldCmpSTDistanceLeft format=(show-scalars,show-types)
+SELECT * FROM geom_geog WHERE st_distance(geog, 'point(0.0 0.0)') < val::int
+----
+select
+ ├── columns: geom:1(geometry) geog:2(geography) val:3(float)
+ ├── immutable
+ ├── scan geom_geog
+ │    └── columns: geom:1(geometry) geog:2(geography) val:3(float)
+ └── filters
+      └── function: st_dwithinexclusive [type=bool, outer=(2,3), immutable]
+           ├── variable: geog:2 [type=geography]
+           ├── const: '0101000020E610000000000000000000000000000000000000' [type=geography]
+           └── cast: FLOAT8 [type=float]
+                └── cast: INT8 [type=int]
+                     └── variable: val:3 [type=float]
+
 # --------------------------------------------------
 # FoldCmpSTDistanceRight
 # --------------------------------------------------

